appmgr/src/logs.rs

```diff
@@ -0,0 +1,178 @@
+use std::ascii::AsciiExt;
+use std::process::Stdio;
+use std::time::{Duration, UNIX_EPOCH};
+
+use anyhow::anyhow;
+use chrono::{DateTime, Utc};
+use clap::ArgMatches;
+use futures::TryStreamExt;
+use rpc_toolkit::command;
+use serde::{Deserialize, Serialize};
+use tokio::io::{AsyncBufReadExt, BufReader};
+use tokio::process::Command;
+use tokio_stream::wrappers::LinesStream;
+
+use crate::context::{EitherContext};
+use crate::error::ResultExt;
+use crate::system::SYSTEMD_UNIT;
+use crate::util::Reversible;
+use crate::Error;
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
+#[serde(rename_all = "kebab-case")]
+pub struct LogResponse {
+    entries: Reversible<LogEntry>,
+    start_cursor: Option<String>,
+    end_cursor: Option<String>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
+pub struct LogEntry {
+    timestamp: DateTime<Utc>,
+    message: String,
+}
+impl std::fmt::Display for LogEntry {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{} {}", self.timestamp, self.message)
+    }
+}
+
+fn display_logs(all: LogResponse, _: &ArgMatches<'_>) {
+    for entry in all.entries.iter() {
+        println!("{}", entry);
+    }
+}
+
+#[command(display(display_logs))]
+pub async fn logs(
+    #[context] _: EitherContext,
+    #[arg] id: String,
+    #[arg] limit: Option<usize>,
+    #[arg] cursor: Option<String>,
+    #[arg] before_flag: Option<bool>,
+) -> Result<LogResponse, Error> {
+    Ok(logs_inner(id, limit, cursor, before_flag.unwrap_or(false)).await?)
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct JournalctlEntry {
+    #[serde(rename = "__REALTIME_TIMESTAMP")]
+    timestamp: String,
+    #[serde(rename = "MESSAGE")]
+    message: String,
+    #[serde(rename = "__CURSOR")]
+    cursor: String,
+}
+impl JournalctlEntry {
+    fn log_entry(self) -> Result<(String, LogEntry), Error> {
+        Ok((
+            self.cursor,
+            LogEntry {
+                timestamp: DateTime::<Utc>::from(
+                    UNIX_EPOCH + Duration::from_micros(self.timestamp.parse::<u64>()?),
+                ),
+                message: self.message,
+            },
+        ))
+    }
+}
+
+async fn logs_inner(
+    id: String,
+    limit: Option<usize>,
+    cursor: Option<String>,
+    before_flag: bool,
+) -> Result<LogResponse, Error> {
+    let limit = limit.unwrap_or(50);
+    let limit_formatted = format!("-n{}", limit);
+    let mut get_prev_logs_and_reverse = false;
+
+    let mut args_out = vec!["--output=json","--output-fields=MESSAGE",&limit_formatted,];
+    let id_formatted = match id.as_str() {
+        SYSTEMD_UNIT=> {
+            args_out.push("-u");
+            SYSTEMD_UNIT.to_owned()
+        },
+        _ => format!("CONTAINER_NAME={}", id)
+    };
```

Comment:
this feels like an ugly hack. Use an enum for the inner fn.

Reply:
will do

Reply:
done